File: presentation/src/main/java/com/mbmc/template/di/ApiModule.java
package com.mbmc.template.di;

import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import com.mbmc.template.data.api.ApiService;
import com.mbmc.template.data.Constants;
import com.mbmc.template.data.api.NetworkInterceptor;
import com.mbmc.template.data.session.SessionManager;
import dagger.Module;
import dagger.Provides;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import javax.inject.Singleton;
import java.util.concurrent.TimeUnit;

@Module
public class ApiModule {

    private static final int TIMEOUT = 10; // 10s

    @Provides
    @Singleton
    Retrofit retrofit(OkHttpClient okHttpClient) {
        return new Retrofit.Builder()
                .baseUrl(Constants.GITHUB_API)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();
    }

    @Provides
    @Singleton
    ApiService apiService(Retrofit retrofit) {
        return retrofit.create(ApiService.class);
    }

    @Provides
    @Singleton
    OkHttpClient okHttpClient(Interceptor interceptor) {
        OkHttpClient.Builder okHttpClientBuilder = new OkHttpClient.Builder();
        okHttpClientBuilder.connectTimeout(TIMEOUT, TimeUnit.SECONDS);
        okHttpClientBuilder.readTimeout(TIMEOUT, TimeUnit.SECONDS);
        okHttpClientBuilder.writeTimeout(TIMEOUT, TimeUnit.SECONDS);
        okHttpClientBuilder.addInterceptor(interceptor);
        return okHttpClientBuilder.build();
    }

    @Provides
    @Singleton
    Interceptor interceptor(SessionManager sessionManager) {
        return new NetworkInterceptor(sessionManager);
    }

}

File: presentation/src/main/kotlin/com/mbmc/template/di/module/ApiModule.kt
package com.mbmc.template.di.module

import com.mbmc.template.data.api.ApiService
import com.mbmc.template.data.Constants
import com.mbmc.template.data.api.NetworkInterceptor
import com.mbmc.template.data.session.SessionManager
import dagger.Module
import dagger.Provides
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory

import javax.inject.Singleton
import java.util.concurrent.TimeUnit

@Module
open class ApiModule {
    companion object {
        private const val TIMEOUT = 10L // 10s
    }

    @Provides
    @Singleton
    fun retrofit(okHttpClient: OkHttpClient): Retrofit = Retrofit.Builder()
        .baseUrl(Constants.GITHUB_API)
        .client(okHttpClient)
        .addConverterFactory(MoshiConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .build()

    @Provides
    @Singleton
    open fun apiService(retrofit: Retrofit): ApiService = retrofit.create(ApiService::class.java)

    @Provides
    @Singleton
    fun okHttpClient(interceptor: Interceptor): OkHttpClient {
        val okHttpClientBuilder = OkHttpClient.Builder()
        okHttpClientBuilder.connectTimeout(TIMEOUT, TimeUnit.SECONDS)
        okHttpClientBuilder.readTimeout(TIMEOUT, TimeUnit.SECONDS)
        okHttpClientBuilder.writeTimeout(TIMEOUT, TimeUnit.SECONDS)
        okHttpClientBuilder.addInterceptor(interceptor)
        return okHttpClientBuilder.build()
    }

    @Provides
    @Singleton
    fun interceptor(sessionManager: SessionManager): Interceptor = NetworkInterceptor(sessionManager)
}
